### PR TITLE
Make purging on startup optional for quota keeper.

### PR DESCRIPTION
**Pull request by SergeyTeplyakov**
If the cache restarts during the build (which is a normal procedure), the cache may decide to start content eviction process at a startup when the cache is full.

Effectively, there is a race condition at startup when one part of the cache reloads the session state and tries to re-pin the content that was used by a previous service instance and another part of the cache tries to evict it. In some cases an important content can be evicted causing the build to fail.

This PR adds mitigation by exposing a configuration option that prevents purging on startup.

### CODE REVIEW

File: Public/Src/Cache/ContentStore/Library/Stores/QuotaManagement/LegacyQuotaKeeper.cs [hunk 1]
```diff
@@ -155,7 +157,7 @@ protected override void DisposeCore()
         public override async Task SyncAsync(Context context)
         {
             // Ensure there are no pending requests.
-            var emptyRequest = new Request<QuotaKeeperRequest, string>(QuotaKeeperRequest.Reserve(0));
+            var emptyRequest = new Request<QuotaKeeperRequest, string>(QuotaKeeperRequest.Synchronize());
```
Comment (JuanCarlosGI) [inline Public/Src/Cache/ContentStore/Library/Stores/QuotaManagement/LegacyQuotaKeeper.cs]: Maybe change variable name?

File: Public/Src/Cache/ContentStore/Library/Stores/FileSystemContentStoreInternal.cs [hunk 2]
```diff
@@ -669,15 +669,18 @@ protected override async Task<BoolResult> StartupCoreAsync(OperationContext cont
                 GetPinnedSize,
                 _nagleQueue);
 
+            var quotaKeeperConfiguration = QuotaKeeperConfiguration.Create(
+                Configuration,
+                _distributedEvictionSettings,
+                size,
+                _settings.StartPurgingAtStartup,
```
Comment (kewilli) [inline Public/Src/Cache/ContentStore/Library/Stores/FileSystemContentStoreInternal.cs]: >_settings [](start = 16, length = 9)

Why not just pass `_settings` here?

File: Public/Src/Cache/ContentStore/Library/Stores/QuotaManagement/LegacyQuotaKeeper.cs [hunk 3]
```diff
@@ -80,9 +79,12 @@ protected override async Task<BoolResult> StartupCoreAsync(OperationContext cont
             _reserveTask = Task.Run(() => Reserve(new Context(context)));
             _taskTracker = new BackgroundTaskTracker(Component, new Context(context));
 
-            // Start purging immediately on startup to clear out residual content in the cache
-            // over the cache quota
-            StartPurging();
+            if (PurgeAtStartup)
```
Comment (kewilli) [inline Public/Src/Cache/ContentStore/Library/Stores/QuotaManagement/LegacyQuotaKeeper.cs]: >PurgeAtStartup [](start = 16, length = 14)

Can we emit some diagnostic when we choose to not purge at startup?

Comment (kewilli) [reply Public/Src/Cache/ContentStore/Library/Stores/QuotaManagement/LegacyQuotaKeeper.cs]: It seems like we would have sufficient logging for when we do purge at startup.

---
In reply to: [273233305](https://github.com/Microsoft/BuildXL/pull/102#discussion_r273233305) [](ancestors = 273233305)

Comment (SergeyTeplyakov) [reply Public/Src/Cache/ContentStore/Library/Stores/QuotaManagement/LegacyQuotaKeeper.cs]: That makes a lot of sense! Will add.

---
In reply to: [273233305](https://github.com/Microsoft/BuildXL/pull/102#discussion_r273233305) [](ancestors = 273233305)

File: Public/Src/Cache/ContentStore/Library/Stores/QuotaManagement/QuotaKeeperConfiguration.cs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        /// <see cref="ContentStoreConfiguration.InitialElasticSize"/>.
+        /// </summary>
+        public MaxSizeQuota InitialElasticSize { get; private set; }
+
+        /// <summary>
+        /// <see cref="ContentStoreConfiguration.HistoryWindowSize"/>.
+        /// </summary>
+        public int? HistoryWindowSize { get; private set; }
+
+        /// <nodoc />
```
Comment (kewilli) [inline Public/Src/Cache/ContentStore/Library/Stores/QuotaManagement/QuotaKeeperConfiguration.cs]: ><nodoc /> [](start = 12, length = 9)

`<see cref="DistributedEvictionSettings"/>` ?

File: Public/Src/Cache/ContentStore/Library/Stores/QuotaManagement/QuotaKeeperV2.cs [hunk 5]
```diff
@@ -108,10 +107,13 @@ protected override async Task<BoolResult> StartupCoreAsync(OperationContext cont
                 () => ProcessReserveRequestsAsync(context.CreateNested()),
                 TaskCreationOptions.LongRunning).Unwrap();
 
-            // Start purging immediately on startup to clear out residual content in the cache
-            // over the cache quota
-            const string Operation = "PurgeRequest";
-            SendPurgeRequest(context, "Startup").FireAndForget(context, Operation);
+            if (PurgeAtStartup)
```
Comment (kewilli) [inline Public/Src/Cache/ContentStore/Library/Stores/QuotaManagement/QuotaKeeperV2.cs]: >PurgeAtStartup [](start = 16, length = 14)

Ditto here

File: Public/Src/Cache/ContentStore/Library/Stores/FileSystemContentStoreInternal.cs (hunk 2 above)
Comment (SergeyTeplyakov) [reply Public/Src/Cache/ContentStore/Library/Stores/FileSystemContentStoreInternal.cs]: God idea.

---
In reply to: [273230438](https://github.com/Microsoft/BuildXL/pull/102#discussion_r273230438) [](ancestors = 273230438)

**Pull request merged**